### PR TITLE
PBP 95481 ZSF VA Email Notification

### PR DESCRIPTION
**Pull request by wayne-weibel**
## Summary

add general functionality to send email notification for a SavedClaim

## Related issue(s)

[ZSF | general claim email notification functionality](https://github.com/department-of-veterans-affairs/va.gov-team/issues/95481)
https://github.com/department-of-veterans-affairs/vsp-infra-application-manifests/pull/3181
https://github.com/department-of-veterans-affairs/vsp-infra-application-manifests/pull/3185

## Testing done

- [x] *New code is covered by unit tests*
- locally in console:
```
3.3.3 :001 > require 'va_notify/notification_email/saved_claim'
3.3.3 :002 > c = Pensions::SavedClaim.find(4)
3.3.3 :003 > vne = VANotify::NotificationEmail::SavedClaim.new(c)
3.3.3 :004 > vne.deliver(:confirmation)
```

## Screenshots
image

## What areas of the site does it impact?
pension and burial at the moment.  any other saved claim form can use this class by extending or directly

## Acceptance criteria

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [x]  Events are being sent to the appropriate logging solution
- [x]  Documentation has been updated (link to documentation)
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [x]  I added a screenshot of the developed feature

## Requested Feedback

Should the `rescue` also reraise the exception?  My thinking in not doing so is that the send of the email should not be a requirement to continue the calling process, and there is the monitoring of success/failure (at least for calling the EmailJob.perform)

### CODE REVIEW

File: lib/va_notify/notification_email.rb [hunk 1]
```diff
@@ -0,0 +1,72 @@
+# frozen_string_literal: true
+
+# Library for VA Notify
+module VANotify
```
Comment (balexandr) [inline lib/va_notify/notification_email.rb:4]: @wayne-weibel with the new monitor classees should this maybe go under `logging/va_notify.rb` so its right there with `monitor` and `sidekiq`?

File: lib/va_notify/notification_email/saved_claim.rb [hunk 2]
```diff
@@ -0,0 +1,155 @@
+# frozen_string_literal: true
+
+require 'va_notify/notification_email'
+
+require 'pry-byebug'
```
Comment (balexandr) [inline lib/va_notify/notification_email/saved_claim.rb:5]: Does this still need to be here?

File: lib/va_notify/notification_email/saved_claim.rb [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        VANotify::NotificationEmail.monitor_send_failure(e&.message, tags:, context:)
+      end
+
+      private
+
+      attr_reader :claim, :email_template_id
+
+      # return or default the service_name to be used
+      def vanotify_service
+        @vanotify_service ||= claim.form_id.downcase.gsub(/-/, '_')
```
Comment (balexandr) [inline lib/va_notify/notification_email/saved_claim.rb:66]: I suspect there will never be nils to cause a problem here so we don't need to:
`claim&.form_id&.downcase&.gsub(/-/, '_')` ?

File: lib/va_notify/notification_email/saved_claim.rb [hunk 4]
```diff
@@ -0,0 +1,155 @@
+# frozen_string_literal: true
+
+require 'va_notify/notification_email'
+
+require 'pry-byebug'
+
+module VANotify
```
Comment (balexandr) [inline lib/va_notify/notification_email/saved_claim.rb:7]: Love this class

File: modules/pensions/app/models/pensions/saved_claim.rb [hunk 5]
```diff
@@ -69,52 +69,12 @@ def email
     end
 
     ##
-    # enqueue the sending of the submission confirmation email
+    # utility function to retrieve claimant first name from form
     #
-    # @see VANotify::EmailJob
+    # @return [String] the claimant first name
     #
-    def send_confirmation_email
```
Comment (balexandr) [inline modules/pensions/app/models/pensions/saved_claim.rb:76]: Yeah this move made sense to do in this ticket

File: lib/va_notify/notification_email/saved_claim.rb (hunk 3 above)
Comment (wayne-weibel) [reply lib/va_notify/notification_email/saved_claim.rb:66]: hmm, i would hope not, but there is a chance bad data could get passed in.  in theory it should be caught during testing (the vanotify.service would not be found and raise an error) but worth safeguarding

File: lib/va_notify/notification_email/saved_claim.rb (hunk 2 above)
Comment (wayne-weibel) [reply lib/va_notify/notification_email/saved_claim.rb:5]: oops, nope.  thanks!

File: modules/pensions/app/models/pensions/saved_claim.rb (hunk 5 above)
Comment (wayne-weibel) [reply modules/pensions/app/models/pensions/saved_claim.rb:76]: yeah, was here or in a separate one specifically for getting the pension/burial emails rigged up

File: lib/va_notify/notification_email.rb (hunk 1 above)
Comment (wayne-weibel) [reply lib/va_notify/notification_email.rb:4]: 50/50 - this monitoring is specific to this module.  can debate and move once the general monitoring changes are in too

**Review by balexandr — DISMISSED** [review]
LGTM

File: .github/CODEOWNERS [hunk 6]
```diff
@@ -1003,6 +1003,7 @@ lib/vic @department-of-veterans-affairs/octo-identity
 lib/virtual_regional_office @department-of-veterans-affairs/va-api-engineers @department-of-veterans-affairs/backend-review-group
 lib/vre @department-of-veterans-affairs/benefits-non-disability @department-of-veterans-affairs/va-api-engineers @department-of-veterans-affairs/backend-review-group
 lib/res @department-of-veterans-affairs/benefits-non-disability @department-of-veterans-affairs/va-api-engineers @department-of-veterans-affairs/backend-review-group
+lib/va_notify @department-of-veterans-affairs/va-api-engineers @department-of-veterans-affairs/backend-review-group
```
Comment (rmtolmach) [inline .github/CODEOWNERS:1006]: These CODWONERS updates don't seem right 😕 Your team will need to own this directory, or you can specify the files that yall specifically own.

File: config/settings.yml [hunk 7]
```diff
@@ -1288,11 +1288,9 @@ vanotify:
       api_key: fake_secret
       template_id:
         appoint_a_representative_confirmation_email: appoint_a_representative_confirmation_email_template_id
-        burial_claim_confirmation_email_template_id: burial_claim_confirmation_email_template_id
         ch31_vbms_form_confirmation_email: ch31_vbms_fake_template_id
         ch31_central_mail_form_confirmation_email: ch31_central_mail_fake_template_id
         career_counseling_confirmation_email: career_counseling_confirmation_email_template_id
-        form527ez_confirmation_email: form527ez_confirmation_email_template_id
         preneeds_burial_form_email: preneeds_burial_form_email_template_id
```
Comment (rmtolmach) [inline config/settings.yml:1294]: Will you also submit a PR to the manifest repo to remove these values from there?

File: config/settings.yml [hunk 8]
```diff
@@ -1399,6 +1397,27 @@ vanotify:
...
+      api_key: fake_secret
+      email:
+        confirmation:
+          template_id: burial_claim_confirmation_email_template_id
+          flipper_id: false
+        error: null
+        received: null
+    burials: *vanotify_services_burial
+    21p_530: *vanotify_services_burial
+    21p_530v2: *vanotify_services_burial
```
Comment (rmtolmach) [inline config/settings.yml:1420]: Will all of these values be added to the manifest repo?

File: spec/lib/va_notify/notification_email/saved_claim_spec.rb [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+      expect(Rails.logger).to receive(:error)
+
+      notification.deliver(:confirmation)
+    end
+
+    it 'records a duplicate attempt' do
+      allow(fake_claim).to receive(:va_notification?).and_return true
+
+      metric = "#{VANotify::NotificationEmail::STATSD}.duplicate_attempt"
+      # expect(VANotify::NotificationEmail).to receive(:monitor_duplicate_attempt)
```
Comment (rmtolmach) [inline spec/lib/va_notify/notification_email/saved_claim_spec.rb:86]: There are a number of commented out lines in this spec file. Is that intentional?

File: config/settings.yml (hunk 7 above)
Comment (wayne-weibel) [reply config/settings.yml:1294]: https://github.com/department-of-veterans-affairs/vsp-infra-application-manifests/pull/3181
https://github.com/department-of-veterans-affairs/vsp-infra-application-manifests/pull/3185

File: config/settings.yml (hunk 8 above)
Comment (wayne-weibel) [reply config/settings.yml:1420]: yes

**Comment by rmtolmach** [comment]
There was a seemingly-unrelated test failure. I'm re-running them for you.

**Comment by rmtolmach** [comment]
Maybe it is related. It keeps failing on your branch and I don't see it failing on `master` https://github.com/department-of-veterans-affairs/vets-api/commits/master/
Here's one. but see the test output for more failures. 
```
  1) SavedClaim::EducationBenefits::VA10203#after_submit calls get_gi_bill_status on the service
     Failure/Error: Settings.vanotify.services.va_gov.template_id.form21_10203_confirmation_email,

     NoMethodError:
       undefined method `template_id' for nil
     # ./app/models/saved_claim/education_benefits/va_10203.rb:121:in `send_confirmation_email'
     # ./app/models/saved_claim/education_benefits/va_10203.rb:22:in `after_submit'
     # ./spec/models/saved_claim/education_benefits/va10203_spec.rb:31:in `block (3 levels) in <top (required)>'
     # /usr/local/bundle/cache/ruby/3.3.0/gems/webmock-3.24.0/lib/webmock/rspec.rb:39:in `block (2 levels) in <top (required)>'
     # /usr/local/bundle/cache/ruby/3.3.0/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:124:in `block in run'
     # /usr/local/bundle/cache/ruby/3.3.0/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:110:in `run'
     # /usr/local/bundle/cache/ruby/3.3.0/gems/rspec-retry-0.6.2/lib/rspec_ext/rspec_ext.rb:12:in `run_with_retry'
     # /usr/local/bundle/cache/ruby/3.3.0/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:37:in `block (2 levels) in setup'
```

**Comment by wayne-weibel** [comment]
> Maybe it is related. It keeps failing on your branch and I don't see it failing on `master` https://github.com/department-of-veterans-affairs/vets-api/commits/master/ Here's one. but see the test output for more failures.

that was a different one from what i was seeing before - i merge master again and will be checking more closely if it fails

**Comment by rmtolmach** [comment]
> > Maybe it is related. It keeps failing on your branch and I don't see it failing on `master` https://github.com/department-of-veterans-affairs/vets-api/commits/master/ Here's one. but see the test output for more failures.
> 
> that was a different one from what i was seeing before - i merged master again and will be checking more closely if it fails

👍 And I re-ran the test suite twice, so it had three passing runs. 

@wayne-weibel Can you get a teammate to review?

**Review by rmtolmach — APPROVED** [review]

**Pull request merged**